_posts/2022-06-01-final-checkpoint-part1.md

```diff
@@ -0,0 +1,173 @@
+then if there are failovers after that (like due to other unfinished tasks get failed), these records would be
+re-emitted and cause data duplication; if we discard these records as now, these records would be lost and cause data loss. 
+
+This issue also exists for the stop-with-savepoint case: the current implementation first stalls the execution and
+takes a savepoint, then it notifies all the sources to start finishing. However, with `–-drain` the operator logic like
+`endInput()` would still be executed during finishing and generate new records, these records would not be committed.
+In addition to that, we could see that in fact the savepoint also does not include these records and thus it is
+inconsistent, which makes the situation worse. 
+
+To provide an opportunity for the operators to commit the last piece of data, we need to modify the process of finishing
+to wait for one more checkpoint after tasks processed all data. An intuitive idea is to directly insert a step to the
+tasks’ lifecycle to wait for the next checkpoint, as shown in the left part of Figure 3. However, it could not solve
+all the issues.
+
+<center>
+<img vspace="20" style="width:90%" src="{{site.baseurl}}/img/blog/2022-06-01-final-checkpoint/finish_cmp.png" />
+<p style="font-size: 0.6em;text-align:left;margin-top:-1em;margin-bottom: 4em">
+  Figure 3. A comparison of the two options to make tasks wait for one more checkpoint before getting finished. The first
+  option directly inserts a step in the tasks’ lifecycle to wait for the next checkpoint, which disallows the tasks to wait
+  for the same checkpoint / savepoint. The second option decouples the notification of all records processed and task finished
+  and allows all tasks to first process all records, then they have the chance to wait for the save checkpoint / savepoint. 
+</p>
+</center>
+
+For the case of bounded sources, the intuitive idea works, but it might have performance issues in some cases:
+as exemplified in Figure 4, If there are multiple cascading tasks containing two-phase commit sinks, each task would
+wait for the next checkpoint separately, thus the job needs to wait for three more checkpoints during finishing,
+which might prolong the total execution time for a long time. 
+
+<center>
+<img vspace="20" style="width:90%" src="{{site.baseurl}}/img/blog/2022-06-01-final-checkpoint/example_job.png" />
+<p style="font-size: 0.6em;text-align:center;margin-top:-1em;margin-bottom: 4em">
+    Figure 4. An example job that contains a chain of tasks containing two-phase-commit operators. 
+</p>
+</center>
+
+For the case of `stop-with-savepoint [--drain]`, the intuitive idea is also flawed since different tasks have to
+wait for different checkpoints / savepoints, thus we could not finish the job with a specific savepoint.
+
+To further overcome these issues, we’d better decouple "finishing operator logic" and "finishing tasks": if we could first
```

Comment:
"finishing operator logic" "finishing tasks", I prefer to use italic or bold here.